SimPEG/electromagnetics/static/utils/static_utils.py

```diff
@@ -258,291 +300,260 @@ def geometric_factor(survey_object, space_type="half space", **kwargs):
-            f"{plot_type} is not recognized"
-        )
+    
 
-    # Set depth to 0 for now
-    z0 = 0.0
-    rho = []
-
-    if not isinstance(data, Data):
-        raise Exception(
-            "A Data instance ({datacls}: <{datapref}.{datacls}>) must be "
-            "provided as the second input. The provided input is a "
-            "{providedcls} <{providedpref}.{providedcls}>".format(
-                datacls=Data.__name__,
-                datapref=Data.__module__,
-                providedcls=data.__name__,
-                providedpref=data.__module__,
-            )
-        )
-    # Use dobs in survey if dobs is None
-    if dobs is None:
-        dobs = data.dobs
-
-    midx, midz = source_receiver_midpoints(data.survey)
-    if midx.shape[1] == 2:
-        min_x, min_y = np.min(midx, axis=0)
-        max_x, max_y = np.max(midx, axis=0)
-        if max_x - min_x > max_y - min_y:
-            midx = midx[:, 0]
-        else:
-            midx = midx[:, 1]
+    # Find all unique line id
+    unique_lineID = np.unique(lineID)
+    
+    # If you output indexing to keep track of possible sorting
+    if output_indexing:
+        k = np.arange(0, survey.nD)
+        out_indices_list = []
+    
+    ab_locs_all = np.c_[survey.a_locations, survey.b_locations]
+    mn_locs_all = np.c_[survey.m_locations, survey.n_locations]
+    
+    # For each unique lineID
+    survey_list = []
+    for ID in unique_lineID:
+        
+        source_list = []
+        
+        # Source locations for this line
+        lineID_index = np.where(lineID == ID)[0]
+        ab_locs, ab_index = np.unique(ab_locs_all[lineID_index, :], axis=0, return_index=True)
+        
+        # Find s=0 location and heading for line
+        start_index = lineID_index[ab_index]
+        if output_indexing:
+            out_indices = []
+            kID = k[lineID_index]                     # data indices part of this line
+        r0 = mkvc(ab_locs_all[start_index[0], 0:2])   # (x0, y0) for the survey line
+        rN = mkvc(ab_locs_all[start_index[-1], 0:2])  # (x, y) for last electrode
+        uvec = (rN - r0)/np.sqrt(np.sum((rN-r0)**2))  # unit vector for line orientation
+        
+        # Along line positions and elevation for electrodes on current line
+        # in terms of position elevation
+        a_locs_s = np.c_[
+            np.dot(ab_locs_all[lineID_index, 0:2]-r0[0], uvec), ab_locs_all[lineID_index, 2]
+        ]
+        b_locs_s = np.c_[
+            np.dot(ab_locs_all[lineID_index, 3:5]-r0[0], uvec), ab_locs_all[lineID_index, -1]
+        ]
+        m_locs_s = np.c_[
+            np.dot(mn_locs_all[lineID_index, 0:2]-r0[0], uvec), mn_locs_all[lineID_index, 2]
+        ]
+        n_locs_s = np.c_[
+            np.dot(mn_locs_all[lineID_index, 3:5]-r0[0], uvec), mn_locs_all[lineID_index, -1]
+        ]
+        
+        # For each source in the line
+        for ii, ind in enumerate(ab_index):
```

Comment:
This could likely call the new `survey_from_ABMN` function (or whatever the actual name is)